File: dsa/concepts/01_arrays/java/L189_RotateArray.java
package li01_array.java;

import java.util.Arrays;

public class L189_RotateArray {
    public static void main(String[] args) {
        L189_RotateArray obj = new L189_RotateArray();

        int nums[] = {1,2,3,4,5,6,7};
        int k = 3;
        obj.rotate(nums, k);
        System.out.println(Arrays.toString(nums));
    }

    // Problem: https://leetcode.com/problems/rotate-array/
    // Idea: Reverse the array, then reverse the first k elements and then reverse the remaining elements.
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k = k % n;
        int l=0, r=n-1;
        while(l<r){
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++; r--;
        }

        l=0; r=k-1;
        while(l<r){
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++; r--;
        }

        l=k; r=n-1;
        while(l<r){
            int temp = nums[l];
            nums[l] = nums[r];
            nums[r] = temp;
            l++; r--;
        }
    }
}

File: dsa/concepts/01_arrays/python/l189_rotate_array.py
"""
L189 - Rotate Array

Problem: https://leetcode.com/problems/rotate-array/
Idea: Reverse entire array, then reverse first k elements, then reverse remaining
Time: O(n)
Space: O(1)
"""

def rotate(nums, k):
    n = len(nums)
    k = k % n  # Handle k > n
    
    # Reverse entire array
    reverse(nums, 0, n - 1)
    # Reverse first k elements
    reverse(nums, 0, k - 1)
    # Reverse remaining elements
    reverse(nums, k, n - 1)


def reverse(nums, start, end):
    while start < end:
        nums[start], nums[end] = nums[end], nums[start]
        start += 1
        end -= 1


if __name__ == "__main__":
    nums = [1, 2, 3, 4, 5, 6, 7]
    k = 3
    print(f"Original: {nums}")
    rotate(nums, k)
    print(f"After rotating by {k}: {nums}")
